src/Storages/ExternalStream/Kafka/Kafka.cpp

```diff
@@ -276,45 +363,185 @@ void Kafka::validateMessageKey(const String & message_key_, IStorage * storage,
 }
 
 /// Validate the topic still exists, specified partitions are still valid etc
-void Kafka::validate(const std::vector<int32_t> & shards_to_query)
+void Kafka::validate()
 {
-    if (shards == 0)
-    {
-        std::scoped_lock lock(shards_mutex);
-        /// Recheck again in case in-parallel query chimes in and init the shards already
-        if (shards == 0)
-        {
-            /// We haven't describe the topic yet
-            auto result = getConsumer()->describe(settings->topic.value);
-            if (result.err != ErrorCodes::OK)
-                throw Exception(ErrorCodes::RESOURCE_NOT_FOUND, "{} topic doesn't exist", settings->topic.value);
+    auto consumer = getConsumer();
+    RdKafka::Topic rkt {*consumer->getHandle(), topic()};
+    auto parition_count = rkt.describe();
+    if (parition_count < 1)
+        throw Exception(ErrorCodes::INVALID_SETTING_VALUE, "Topic has no paritions, topic={}", topic());
 
-            shards = result.partitions;
-        }
-    }
-
-    if (!shards_to_query.empty())
+    if (!shards_from_settings.empty())
     {
         /// User specified specific partitions to consume.
         /// Make sure they are valid.
-        for (auto shard : shards_to_query)
+        for (auto shard : shards_from_settings)
         {
-            if (shard >= shards)
+            if (shard >= parition_count)
                 throw Exception(
                     ErrorCodes::INVALID_SETTING_VALUE,
-                    "Invalid topic partition {} for topic {}, biggest partition is {}",
+                    "Invalid topic partition {} for topic {}, total number of partitions is {}",
                     shard,
                     settings->topic.value,
-                    shards);
+                    parition_count);
         }
     }
 }
 
+Pipe Kafka::read(
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t max_block_size,
+    size_t /*num_streams*/)
+{
+    auto consumer = getConsumer();
+    /// The topic_ptr can be shared between all the sources in the same pipe, because each source reads from a different partition.
+    auto topic_ptr = std::make_shared<RdKafka::Topic>(*consumer->getHandle(), topic());
```

Comment:
Yes, this is how `rd_kafka_consume_start` works. The same `topic` can't be used at the same time in multple select queries. Not just the topic, a `consumer` can't be used in multple select queries at the same time either, that's why we need a consumer pool here.

Reply:
in the ctor of `RdKafka::Topic`, we always call `rd_kafka_topic_new`, it's behavior is unreasonable

Reply:
I c, it's okay
```c++
 * Topic handles are refcounted internally and calling rd_kafka_topic_new()
 * again with the same topic name will return the previous topic handle
 * without updating the original handle's configuration.
 * Applications must eventually call rd_kafka_topic_destroy() for each
 * succesfull call to rd_kafka_topic_new() to clear up resources.
```

Reply:
Also, because this is in the same external stream, the configuration will always be the same. Hence, `return the previous topic handle without updating the original handle's configuration` is totally fine.